README.rst
trio_redis
==========

A Redis client for `Trio`_

.. _Trio: https://trio.readthedocs.io/en/stable/

Examples
--------

See the tests for examples.

Status
------

- Before v1.0.0 the API is not stable.
- Alpha quality, please test carefully before use.
- A large part of the Redis commands is not implemented yet.


changelog.md
# Changelog

## 0.3.0

- Get db from Redis URL path part in _redis._parse_url().
- Use a local Redis server for testing. Tests are much faster.
- Add RedisSentinel, a client that gets the address of the Redis server
  from a list Sentinel servers (on connect and on reconnect after a
  fail-over).
- RedisPool now accepts a factory function to allow pooling of any
  Redis-like client (e.g. Redis and RedisSentinel).

## 0.2.3

- Replace Poetry with Setuptools.


## 0.2.2

- Reset connection (reconnect with Redis) after a call to execute()
  or execute_many() has been cancelled.
- Add xread command.


## 0.2.1

- Remove setup.py shim.


## 0.2.0

- Add API for Redis streams commands:
  - xpending
  - xinfo_groups
  - xinfo_stream
  - xdel
  - xtrim
- Return an empty list when an empty pipeline is awaited.
- Raise ClosedError when connection is unexpectedly closed.
- Use Poetry for dependency management and packaging.


tests/conftest.py
from trio_redis.testing_utils import (  # noqa: F401
    pytest_addoption,
    redis,
    redis_pool,
    redis_sentinel_cluster,
    redis_server,
    redis_url,
)


tests/test_redis.py
import pytest
import trio

from trio_redis import Redis
from trio_redis._redis import Pipeline, _parse_url
from trio_redis.testing_utils import TCPProxy


async def test_concurrent_use_of_single_instance(redis):
    # Concurrent use of the same instance is not allowed, because
    # it'll possibly mess up the state of the instance.
    #
    # Why not use a lock and wait for the connection to become available?
    # There's no gain in waiting. Using a pipeline or the pool to execute
    # commands concurrently has more benefits (i.e. it's faster).
    #
    # By raising an exception it should become clear the way the
    # developer is doing things is not the right way.
    with pytest.raises(redis.BusyError):
        async with trio.open_nursery() as nursery:
            nursery.start_soon(redis.set, 'x', 1)
            nursery.start_soon(redis.set, 'y', 1)

    # Sequential use is allowed.
    result = await redis.set('x', 1)
    assert result
    result = await redis.set('y', 1)
    assert result


async def test_double_connect(redis):
    with pytest.raises(redis.BusyError):
        await redis.connect()


async def test_double_aclose(redis_url):
    redis = Redis.from_url(redis_url)
    await redis.connect()
    await redis.aclose()

    with pytest.raises(redis.ClosedError):
        await redis.aclose()


async def test_pipeline(redis):
    pipeline = (redis.pipeline()
        .set('x', 'y')
        .get('x')
        .set('y', 'z')
        .get('y')
        .set('z', 'a')
        .get('z')
    )
    assert isinstance(pipeline, Pipeline)

    result = await pipeline
    assert result == [True, b'y', True, b'z', True, b'a']


async def test_redis_pool_acquire_release(redis_pool):
    c1 = await redis_pool.acquire()
    assert c1 not in redis_pool._free
    assert c1 in redis_pool._not_free

    c2 = await redis_pool.acquire()
    assert c2 not in redis_pool._free
    assert c2 in redis_pool._not_free

    await redis_pool.release(c1)
    assert c1 in redis_pool._free
    assert c1 not in redis_pool._not_free

    await redis_pool.release(c2)
    assert c2 in redis_pool._free
    assert c2 not in redis_pool._not_free


async def test_redis_pool_commands(redis_pool):
    result = await redis_pool.set('x', 1)
    assert result
    result = await redis_pool.get('x')
    assert result == b'1'


async def test_redis_pool_borrow(redis_pool):
    async with redis_pool.borrow() as client:
        await client.set('a', 1)


async def test_close_socket(nursery, redis_url):
    class ProxyUnexpectedClose(TCPProxy):
        async def handle_request(self, request):
            if b'GET' in request:
                raise self.CloseClientConnection
            return request

    proxy = ProxyUnexpectedClose(redis_url)
    address = await nursery.start(proxy.run_forever)

    redis = Redis(*address)
    await redis.connect()

    with pytest.raises(redis.ClosedError, match='connection unexpectedly closed'):
        await redis.get('x')

    await redis.aclose()
    nursery.cancel_scope.cancel()


async def test_connection_reset_after_cancelled(redis, redis_url):
    # NOTE: The redis fixture is needed to cleanup after this test ends.

    reader = Redis.from_url(redis_url)
    await reader.connect()
    writer = Redis.from_url(redis_url)
    await writer.connect()

    async def read():
        with trio.move_on_after(0.5) as cancel_scope:
            await reader.xread({'x': '0-0'}, block=5000)

        # Assert that XREAD has been cancelled.
        assert cancel_scope.cancelled_caught

        # Sleep for a bit. In the meanwhile an XADD is executed.
        await trio.sleep(0.5)

        # Try to get a value from a non-existing key.
        # If the connection was not reset we'd get the message which
        # was added by XADD in the write() function.
        result = await reader.get('bleep')
        assert result is None

    async def write():
        # Sleep until the XREAD is cancelled, then do XADD.
        await trio.sleep(0.8)
        await writer.xadd('x', {'a': 1})

    async with trio.open_nursery() as nursery:
        nursery.start_soon(read)
        nursery.start_soon(write)


@pytest.mark.parametrize('input_url,expected_kwargs', [
    ('redis://localhost', {'host': 'localhost', 'port': 6379}),
    ('redis://:26379', {'host': 'localhost', 'port': 26379}),
    ('redis://', {'host': 'localhost', 'port': 6379}),
    ('redis://localhost:26379', {'host': 'localhost', 'port': 26379}),
    ('redis://localhost:26379/1', {'host': 'localhost', 'port': 26379, 'db': '1'}),
    ('redis://localhost:26379/', {'host': 'localhost', 'port': 26379}),
    (
        'redis://localhost:26379?foo=bar&qux=baz',
        {'host': 'localhost', 'port': 26379, 'foo': 'bar', 'qux': 'baz'},
    ),
    # NOTE: Confirm that username and password are not supported yet.
    (
        'redis://user:secret@localhost:26379?foo=bar&qux=baz',
        {'host': 'localhost', 'port': 26379, 'foo': 'bar', 'qux': 'baz'},
    ),
])
def test_parse_url(input_url, expected_kwargs):
    assert _parse_url(input_url) == expected_kwargs


@pytest.mark.parametrize('input_url,partial_error', [
    ('://localhost', 'missing scheme'),
    ('redis+tls://localhost', 'unsupported scheme'),
    ('redis://localhost/foo', 'db must be a digit'),
    ('redis://localhost/1?foo=bar&foo=bar', 'parameter foo set multiple times'),
])
def test_parse_invalid_url(input_url, partial_error):
    with pytest.raises(ValueError) as excinfo:
        _parse_url(input_url)
    assert partial_error in str(excinfo.value)


tests/test_cluster.py
from trio_redis import RedisCluster


async def test_cluster(redis_cluster):
    redis = RedisCluster(redis_cluster.nodes())
    await redis.connect()
    await redis.aclose()


async def test_cluster_asking(redis_cluster):
    # TODO:
    # - Choose key and get slot.
    # - Set slot to MIGRATING on node A, and IMPORTING on node B.
    #       https://redis.io/commands/cluster-setslot
    # - Exec command on key, expect ASK error to be handled.

    redis = RedisCluster(redis_cluster.nodes())
    await redis.connect()
    await redis.aclose()


trio_redis/_commands.py
# Lots of checking and parsing code is from redis-py.
# See: https://github.com/andymccurdy/redis-py/blob/master/redis/client.py


from functools import partial


__all__ = [
    'ConnectionCommands',
    'KeysCommands',
    'ScriptingCommands',
    'SentinelCommands',
    'ServerCommands',
    'SortedSetCommands',
    'StreamCommands',
    'StringCommands',
]


_EMPTY_TUPLE = tuple()  # noqa: C408


def _to_str(b):
    return b.decode('utf-8')


def bool_ok(reply):
    return reply == b'OK'


def parse_int_bool_list(reply):
    return [r == 1 for r in reply]


def parse_xread(reply):
    if reply is None:
        return []
    return [[r[0], parse_stream_list(r[1])] for r in reply]


def parse_stream_list(reply):
    if reply is None:
        return None
    data = []
    for r in reply:
        if r is not None:
            data.append((r[0], pairs_to_dict(r[1])))
        else:
            data.append((None, None))
    return data


def parse_xinfo_stream(response):
    data = pairs_to_dict(response)

    for k in (b'first-entry', b'last-entry'):
        item = data[k]
        if item is not None:
            data[k] = (item[0], pairs_to_dict(item[1]))

    return data


def pairs_to_dict(reply):
    it = iter(reply)
    return dict(zip(it, it))


def parse_list_of_dicts(reply):
    return [pairs_to_dict(r) for r in reply]


def parse_xpending(response, **options):
    consumers = [{b'name': n, b'pending': p} for n, p in response[3] or []]
    return {
        b'pending': response[0],
        b'min': response[1],
        b'max': response[2],
        b'consumers': consumers
    }


def parse_xpending_range(reply):
    if not reply:
        return []

    return [
        {
            'message_id': r[0],
            'consumer': r[1],
            'time_since_delivered': int(r[2]),
            'times_delivered': int(r[3]),
        }
        for r in reply
    ]


def parse_zadd(reply, as_score=False):
    if reply is None:
        return None
    if as_score:
        return float(reply)
    return int(reply)


def parse_master_addr(reply):
    return (_to_str(reply[0]), int(reply[1]))


class ConnectionCommands:
    def select(self, index):
        return self.execute([b'SELECT', index])


class KeysCommands:
    def exists(self, *keys):
        return self.execute([b'EXISTS', *keys])

    def pttl(self, key):
        return self.execute([b'PTTL', key])

    def keys(self, pattern='*'):
        return self.execute([b'KEYS', pattern])


class ScriptingCommands:
    def eval(self, script, keys=[], args=[]):
        return self.execute([b'EVAL', script, len(keys), *keys, *args])

    def evalsha(self, sha1, keys=[], args=[]):
        return self.execute([b'EVALSHA', sha1, len(keys), *keys, *args])

    def script_load(self, script):
        return self.execute([b'SCRIPT', b'LOAD', script])

    def script_flush(self):
        return self.execute([b'SCRIPT', b'FLUSH'])

    def script_exists(self, *sha1):
        return self.execute([b'SCRIPT', b'EXISTS', *sha1], parse_int_bool_list)


class SentinelCommands:
    """Commands for Redis Sentinel nodes.

    See: https://redis.io/topics/sentinel#sentinel-commands
    """
    def get_master_addr_by_name(self, master_name):
        return self.execute(
            [b'SENTINEL', b'GET-MASTER-ADDR-BY-NAME', master_name],
            parse_master_addr,
        )

    def master(self, master_name):
        return self.execute([b'SENTINEL', b'MASTER', master_name], pairs_to_dict)

    def masters(self):
        return self.execute([b'SENTINEL', b'MASTERS'])

    def myid(self):
        return self.execute([b'SENTINEL', b'MYID'])

    def replicas(self, master_name):
        return self.execute([b'SENTINEL', b'REPLICAS', master_name])

    def sentinels(self, master_name):
        return self.execute([b'SENTINEL', b'SENTINELS', master_name])


class ServerCommands:
    def flushdb(self):
        return self.execute((b'FLUSHDB',))


class SortedSetCommands:
    def zadd(self, key, mapping, nx=False, xx=False, ch=False, incr=False):
        if not mapping:
            raise ValueError('ZADD requires at least one element/score pair')
        if nx and xx:
            raise ValueError("ZADD allows either 'nx' or 'xx', not both")
        if incr and len(mapping) != 1:
            raise ValueError("ZADD option 'incr' only works when passing a"
                             'single element/score pair')

        pieces = [b'ZADD', key]
        as_score = False

        if nx:
            pieces.append(b'NX')
        if xx:
            pieces.append(b'XX')
        if ch:
            pieces.append(b'CH')
        if incr:
            pieces.append(b'INCR')
            as_score = True

        # ((<member>, <score>), …) → ((<score>, <member>), …)
        for pair in mapping:
            pieces.extend([pair[1], pair[0]])

        return self.execute(pieces, partial(parse_zadd, as_score=as_score))

    def zcount(self, key, min='-inf', max='+inf'):
        return self.execute([b'ZCOUNT', key, min, max])


class StreamCommands:
    def xadd(self, key, fields, id=u'*', maxlen=None, approximate=True):
        pieces = [b'XADD', key]

        if maxlen is not None:
            if not isinstance(maxlen, int) or maxlen < 1:
                raise ValueError('XADD maxlen must be a positive integer')
            pieces.append(b'MAXLEN')
            if approximate:
                pieces.append(b'~')
            pieces.append(maxlen)

        pieces.append(id)

        if not isinstance(fields, dict) or len(fields) == 0:
            raise ValueError('XADD fields must be a non-empty dict')
        for pair in fields.items():
            pieces.extend(pair)

        return self.execute(pieces)

    def xread(self, streams, count=None, block=None):
        pieces = [b'XREAD']

        if count is not None:
            if not isinstance(count, int) or count < 1:
                raise ValueError('XREADGROUP count must be a positive integer')
            pieces.extend([b'COUNT', count])

        if block is not None:
            if not isinstance(block, int) or block < 0:
                raise ValueError('XREADGROUP block must be a non-negative '
                                 'integer')
            pieces.extend([b'BLOCK', block])

        pieces.append(b'STREAMS')
        pieces.extend(streams.keys())
        pieces.extend(streams.values())

        return self.execute(pieces, parse_xread)

    def xreadgroup(
        self,
        groupname,
        consumername,
        streams,
        count=None,
        block=None,
    ):
        pieces = [b'XREADGROUP', b'GROUP', groupname, consumername]

        if count is not None:
            if not isinstance(count, int) or count < 1:
                raise ValueError('XREADGROUP count must be a positive integer')
            pieces.extend([b'COUNT', count])

        if block is not None:
            if not isinstance(block, int) or block < 0:
                raise ValueError('XREADGROUP block must be a non-negative '
                                 'integer')
            pieces.extend([b'BLOCK', block])

        if not isinstance(streams, dict) or not streams:
            raise ValueError('XREADGROUP streams must be a non empty dict')

        pieces.append(b'STREAMS')
        pieces.extend(streams.keys())
        pieces.extend(streams.values())

        return self.execute(pieces, parse_xread)

    def xgroup_create(self, key, groupname, id=b'$', mkstream=False):
        pieces = [b'XGROUP', b'CREATE', key, groupname, id]
        if mkstream:
            pieces.append(b'MKSTREAM')
        return self.execute(pieces, bool_ok)

    def xack(self, key, groupname, *ids):
        return self.execute([b'XACK', key, groupname, *ids], int)

    def xpending(self, key, groupname):
        return self.execute([b'XPENDING', key, groupname], parse_xpending)

    def xpending_range(self, key, groupname, start, end, count, consumername=None):
        # NOTE: The start, end and count arguments are optional to
        # XPENDING, but we only use the command with these arguments.
        # So they are not optional here.

        pieces = [b'XPENDING', key, groupname, start, end, count]
        if consumername is not None:
            pieces.append(consumername)
        return self.execute(pieces, parse_xpending_range)

    def xclaim(
        self,
        key,
        groupname,
        consumername,
        min_idle_time,  # Milliseconds.
        message_ids,
    ):
        # NOTE: XCLAIM is partially implemented. The optional arguments
        # are not used and there's no benefit in implementing things
        # that are not used.

        if not isinstance(min_idle_time, int) or min_idle_time < 0:
            raise ValueError('XCLAIM min_idle_time must be a non negative '
                             'integer')
        if not isinstance(message_ids, (list, tuple)) or not message_ids:
            raise ValueError('XCLAIM message_ids must be a non empty list or '
                             'tuple of message IDs to claim')

        return self.execute([
            b'XCLAIM',
            key,
            groupname,
            consumername,
            min_idle_time,
            *message_ids
        ], parse_stream_list)

    def xlen(self, key):
        return self.execute([b'XLEN', key], int)

    def xrange(self, key, min='-', max='+', count=None):
        pieces = []
        if count is not None:
            if not isinstance(count, int) or count < 1:
                raise ValueError('XRANGE count must be a positive integer')
            pieces.extend([b'COUNT', str(count)])
        return self.execute([b'XRANGE', key, min, max, *pieces], parse_stream_list)

    def xinfo_groups(self, key):
        return self.execute([b'XINFO', b'GROUPS', key], parse_list_of_dicts)

    def xinfo_stream(self, key):
        return self.execute([b'XINFO', b'STREAM', key], parse_xinfo_stream)

    def xdel(self, key, *ids):
        return self.execute([b'XDEL', key, *ids], int)

    def xtrim(self, key, maxlen, approximate=True):
        pieces = [b'XTRIM', key, b'MAXLEN']

        if approximate:
            pieces.append(b'~')

        pieces.append(maxlen)
        return self.execute(pieces, int)


class StringCommands:
    def set(self, key, value, ex=None, px=None, nx=False, xx=False):
        pieces = [b'SET', key, value]

        if ex is not None:
            pieces.extend([b'EX', ex])
        if px is not None:
            pieces.extend([b'PX', px])

        if nx:
            pieces.append(b'NX')
        if xx:
            pieces.append(b'XX')

        return self.execute(pieces, bool_ok)

    def get(self, key):
        return self.execute([b'GET', key])


trio_redis/_connection.py
import hiredis
import trio

from ._errors import BusyError, ClosedError


__all__ = [
    'Connection',
]


class Connection:
    """A TCP connection to a Redis server."""
    def __init__(self, host, port):
        self.host = host
        self.port = port

        self._is_connected = False
        self._is_busy = False
        self._stream = None
        self._parser = hiredis.Reader()

        self._cleanup_timeout = 5.0

    def __repr__(self):
        return f'<{self.__class__.__name__}: {self.host}:{self.port}>'

    async def connect(self):
        if self._is_connected:
            raise BusyError('already connected')
        self._stream = await trio.open_tcp_stream(self.host, self.port)
        self._is_connected = True

    async def aclose(self):
        if not self._is_connected:
            raise ClosedError('already closed')
        try:
            await self._stream.aclose()
        finally:
            self._stream = None
            self._is_connected = False

    async def execute(self, command):
        return (await self.execute_many((command,)))[0]

    async def execute_many(self, commands):
        if not self._is_connected:
            raise ClosedError('cannot execute command, connection is closed')
        if self._is_busy:
            raise BusyError('another task is currently executing a command')
        self._is_busy = True

        try:
            request = b''.join([_build_request(cmd) for cmd in commands])
            await self._stream.send_all(request)
            reply = await self._read_reply(expected=len(commands))
            return reply
        except trio.Cancelled:
            # If cancelled, reconnect. The old connection might still
            # receive a reply from Redis and if this connection is
            # reused it'll get the reply from the previous command.
            with trio.move_on_after(self._cleanup_timeout) as cleanup_scope:
                cleanup_scope.shield = True
                await self.aclose()
                await self.connect()
            raise
        finally:
            self._is_busy = False

    async def _read_reply(self, expected=1):
        """Read and parse replies from connection.

        ``expected`` is the amount of expected replies. In case of
        pipelining this number is set to the amount of commands sent.
        """
        replies = []

        while True:
            data = await self._stream.receive_some()
            if data == b'':
                raise ClosedError('connection unexpectedly closed')
            self._parser.feed(data)
            while True:
                reply = self._parser.gets()
                if reply is False:
                    break  # Read more data, go back to the outer loop.
                replies.append(reply)
                if len(replies) == expected:
                    return replies


def _build_request(args):
    """Build a RESP request.

    A request consists of a RESP array with RESP bulk strings::

        * [ARRAY LENGTH] \r\n [ARRAY ELEMENTS]

    ``ARRAY LENGTH`` is the length of the array, the number of bulk
    strings. And ``ARRAY ELEMENTS`` contains the bulk strings::

        $ [STRING LENGTH] \r\n [STRING DATA] \r\n

    Here's a real example of ``GET mykey``::

        *2\r\n$3\r\nGET\r\n$5\r\nmykey\r\n

    See `Redis Protocol specification`_ for more information.

    .. _Redis Protocol specification: https://redis.io/topics/protocol

    The hiredis library only exposes and API for decoding responses,
    not building request strings. That's why this function is needed.
    """
    out = b'*%d\r\n' % len(args)

    for part in args:
        if isinstance(part, str):
            part = _str_to_bytes(part)
        elif isinstance(part, int):
            part = b'%d' % part
        elif not isinstance(part, bytes):
            raise ValueError('only bytes, str, and int are supported')
        out += b'$%d\r\n%b\r\n' % (len(part), part)

    return out


def _str_to_bytes(v):
    return bytes(v, encoding='utf-8')


trio_redis/_redis.py
from contextlib import asynccontextmanager
from itertools import repeat
from urllib.parse import urlparse, parse_qs

import hiredis
import trio

from . import _errors
from ._commands import (
    ConnectionCommands,
    KeysCommands,
    ScriptingCommands,
    SentinelCommands,
    ServerCommands,
    SortedSetCommands,
    StreamCommands,
    StringCommands,
)
from ._connection import Connection


__all__ = [
    'Redis',
    'RedisPool',
    'RedisSentinel',
]


_commands = (
    ConnectionCommands,
    KeysCommands,
    ScriptingCommands,
    ServerCommands,
    SortedSetCommands,
    StreamCommands,
    StringCommands,
)


DEFAULT_HOST = 'localhost'
DEFAULT_PORT = 6379


class _BaseRedis:
    """Abstract class for Redis clients."""
    # Alias exceptions. No need to import them. :)
    ConnectError = _errors.ConnectError
    BusyError = _errors.BusyError
    ClosedError = _errors.ClosedError
    ReplyError = _errors.ReplyError
    ReadOnlyError = _errors.ReadOnlyError

    def pipeline(self):
        return Pipeline(self)

    async def execute(self, command, parse_callback=None):
        return (await self.execute_many([command], [parse_callback or _noop]))[0]

    async def execute_many(self, commands, parse_callbacks=None):
        raise NotImplementedError('please implement this')


class _BareRedis(_BaseRedis):
    """Redis client w/o command methods."""

    @classmethod
    def from_url(cls, url):
        kwargs = _parse_url(url)
        return cls(**kwargs)

    def __init__(self, host=None, port=None, db=None):
        self.host = host or DEFAULT_HOST
        self.port = DEFAULT_PORT if port is None else port
        self.db = db
        self._conn = Connection(self.host, self.port)

    async def connect(self):
        await self._conn.connect()
        if self.db is not None:
            await self.select(self.db)

    async def aclose(self):
        await self._conn.aclose()

    async def execute_many(self, commands, parse_callbacks=None):
        if not parse_callbacks:
            parse_callbacks = repeat(_noop)

        replies = await self._conn.execute_many(commands)
        replies = [
            self._parse_reply(reply, cb)
            for reply, cb in zip(replies, parse_callbacks)
        ]

        return replies

    def _parse_reply(self, reply, parse_callback):
        if isinstance(reply, hiredis.ReplyError):
            raise _errors.create_error_from_reply(reply)
        else:
            reply = parse_callback(reply)
        return reply


class RedisPool(_BaseRedis, *_commands):
    """A pool of Redis clients.

    It's not needed to explicitly borrow a client from the pool. All
    commands (except SELECT) are implemented, acquiring and releasing
    clients is done behind the scenes.

    ``client_factory`` is a function that returns a new instance of a
    client class. E.g. an instance of Redis or RedisSentinel.

    ``minimum`` is the minimum amount of clients created. When needed
    new client instances are created until ``maximum`` is reached.

    An instance of this class can be used concurrently. Instances of
    ``Redis`` cannot.
    """
    def __init__(self, client_factory, minimum=1, maximum=10):
        self.client_factory = client_factory
        self.minimum = minimum
        self.maximum = maximum

        self._limit = trio.Semaphore(maximum)
        self._free = []
        self._not_free = []

    @classmethod
    def from_url(cls, client_factory, url):
        kwargs = _parse_url(url)
        return cls(client_factory, **kwargs)

    @classmethod
    def redis(cls, host=None, port=None, db=None, url=None, **pool_kwargs):
        if bool(host or port) == bool(url):
            raise ValueError('either host and port OR url must be given')

        def redis_factory():
            if host or port:
                return Redis(host, port, db)
            else:
                return Redis.from_url(url)

        return cls(redis_factory, **pool_kwargs)

    @classmethod
    def redis_sentinel(cls, master_name, addresses=None, urls=None, **pool_kwargs):
        if bool(addresses) == bool(urls):
            raise ValueError('either addresses OR urls must be given')

        def redis_sentinel_factory():
            if addresses:
                return RedisSentinel(master_name, addresses)
            else:
                return RedisSentinel.from_url(master_name, urls)

        return cls(redis_sentinel_factory, **pool_kwargs)

    async def connect(self):
        async def add():
            self._free.append(await self._new_client())

        async with trio.open_nursery() as nursery:
            for n in range(self.minimum):
                nursery.start_soon(add)

    async def _new_client(self):
        client = self.client_factory()
        await client.connect()
        return client

    async def aclose(self):
        async with trio.open_nursery() as nursery:
            for client in self._free:
                nursery.start_soon(client.aclose)
            for client in self._not_free:
                nursery.start_soon(client.aclose)

        self._free = []
        self._not_free = []

    async def acquire(self):
        """Acquire a client from the pool.

        Blocks if no client if available.
        """
        await self._limit.acquire()

        try:
            client = self._free.pop()
        except IndexError:
            client = await self._new_client()

        self._not_free.append(client)

        return client

    async def release(self, client):
        self._not_free.remove(client)
        self._free.append(client)
        self._limit.release()

    @asynccontextmanager
    async def borrow(self):
        """Explicitly borrow a client from the pool.

        For example::

            async with pool.borrow() as client:
                await client.set('a', 1)
                await client.set('b', 2)
                await client.set('c', 3)
        """
        client = await self.acquire()
        try:
            yield client
        finally:
            await self.release(client)

    async def execute_many(self, commands, parse_callbacks=None):
        client = await self.acquire()
        try:
            return await client.execute_many(commands, parse_callbacks)
        finally:
            await self.release(client)

    def select(self, index):
        raise NotImplementedError('SELECT not implemented for RedisPool')


class Redis(_BareRedis, *_commands):
    """Basic Redis client."""


class RedisSentinel(Redis):
    """Basic Redis client with Sentinel support."""
    RECONNECT_DELAY = 0.5  # Seconds.
    RECONNECT_DELAY_INCREASE = 1.5  # Multiplier.
    MAX_RECONNECT_TRIES = 10

    @classmethod
    def from_url(cls, master_name, urls):
        adresses = []

        for url in urls:
            tmp = _parse_url(url)
            adresses.append((tmp['host'], tmp['port']))

        return cls(master_name, adresses)

    def __init__(self, master_name, adresses):
        self.master_name = master_name
        self.sentinel = Sentinel(adresses)
        super().__init__()

    async def connect(self):
        addr = await self._get_current_master_addr()
        self.host = self._conn.host = addr[0]
        self.port = self._conn.port = addr[1]
        await super().connect()

    async def _reconnect(self):
        if self._conn._is_connected:
            await self._conn.aclose()
        self._conn = Connection(None, None)
        await self.connect()

    async def _get_current_master_addr(self):
        await self.sentinel.connect()
        try:
            return await self.sentinel.get_master_addr_by_name(self.master_name)
        finally:
            await self.sentinel.aclose()

    async def execute_many(self, commands, parse_callbacks=None):
        delay = self.RECONNECT_DELAY
        do_reconnect = False

        for n in range(self.MAX_RECONNECT_TRIES + 1):
            if do_reconnect:
                try:
                    await self._reconnect()
                except OSError:
                    await trio.sleep(delay)
                    delay *= self.RECONNECT_DELAY_INCREASE
                    continue

            try:
                return await super().execute_many(commands, parse_callbacks)
            except (self.ClosedError, self.ReadOnlyError):
                do_reconnect = True

        raise self.ConnectError(f'unable to connect; conn={self._conn}')


class Sentinel(_BareRedis, SentinelCommands):
    """Sentinel client."""
    CONNECT_TIMEOUT = 0.5

    @classmethod
    def from_url(cls, urls):
        adresses = []

        for url in urls:
            tmp = _parse_url(url)
            if 'db' in tmp:
                raise ValueError('sentinal client does not support db selection')
            adresses.append((tmp['host'], tmp['port']))

        return cls(adresses)

    def __init__(self, addresses):
        self._connections = [Connection(*addr) for addr in addresses]
        self._conn = None

    async def connect(self):
        # Behavior:
        # - Try to connect to a sentinel node.
        # - Abort if connect does not happen within CONNECT_TIMEOUT.
        # - Connection errors are handled by catching OSError.
        # - On success, move the connection to the beginning of the
        #   connection list and stop break loop.
        #
        # See: https://redis.io/topics/sentinel-clients
        if self._conn:
            raise self.BusyError('sentinel client already connected')
        for conn in self._connections[:]:
            with trio.move_on_after(self.CONNECT_TIMEOUT) as cancel_scope:
                try:
                    await conn.connect()
                except OSError:
                    continue

            if not cancel_scope.cancelled_caught:
                self._connections.remove(conn)
                self._connections.insert(0, conn)
                self._conn = conn
                return
            else:
                await conn.aclose()

        raise self.ConnectError('unable to connect to sentinel node')

    async def aclose(self):
        await super().aclose()
        self._conn = None


class Pipeline(*_commands):
    """Command pipeliner.

    A pipeline object batches commands and executes all commands when
    the object is awaited.

    An in-order list of replies is returned. If there are no batched
    commands, an empty list is returned.
    """
    def __init__(self, redis):
        self._redis = redis
        self._buffer = []
        self._callbacks = []

    def execute(self, command, parse_callback=None):
        if not parse_callback:
            parse_callback = _noop
        self._buffer.append(command)
        self._callbacks.append(parse_callback)
        return self

    def __await__(self):
        if not self._buffer:
            async def return_empty_list():
                return []
            return return_empty_list().__await__()

        return self._redis.execute_many(
            self._buffer,
            self._callbacks,
        ).__await__()


def _parse_url(url):
    """Parse a Redis URL.

    This is only a partial implementation that only supports a hostname,
    port, and database. See the `URI scheme`_ at IANA for a complete
    description of the URI scheme.

    .. _URI scheme: https://www.iana.org/assignments/uri-schemes/prov/redis
    """
    kwargs = {
        'host': DEFAULT_HOST,
        'port': DEFAULT_PORT,
    }

    url = urlparse(url)
    if not url.scheme:
        raise ValueError(f'missing scheme in {url!r}')
    if url.scheme != 'redis':
        raise ValueError(f'unsupported scheme in {url!r}')

    if url.hostname:
        kwargs['host'] = url.hostname
    if url.port:
        kwargs['port'] = url.port

    db = url.path[1:]
    if db:
        if not db.isdigit():
            raise ValueError(f'db must be a digit in {url!r}')
        kwargs['db'] = db

    params = parse_qs(url.query)
    for k, v in params.items():
        if len(v) > 1:
            raise ValueError(f'parameter {k} set multiple times in {url!r}')
        params[k] = v[0]

    kwargs.update(params)

    return kwargs


def _noop(value):
    return value
